mock/channels/channelService.js

```diff
@@ -1,40 +1,138 @@
+          );
+          break;
+        }
+      }
     }
 
-    async addContextListener(contextType, channel) {
-        await channel.addContextListener(
-            contextType,
-            () => closeFinsembleWindow());
+    //close ChannelsApp when test is complete
+    await this.closeWindowOnCompletion(channel, config);
+
+    //notify app A that ChannelsApp has finished executing
+    if (config.notifyAppAOnCompletion) {
+      await this.notifyAppAOnCompletion(channel, config);
     }
-}
+  }
+
+  async joinSystemChannelOne() {
+    const channels = await window.fdc3.getSystemChannels();
+    await window.fdc3.joinChannel(channels[0].id);
+    return channels[0];
+  }
+
+  //retrieve/create "test-channel" app channel
+  async retrieveTestAppChannel() {
+    return await window.fdc3.getOrCreateChannel("test-channel");
+  }
 
-class UserChannelService {
-    async joinChannel() {
-        const channels = await window.fdc3.getSystemChannels();
-        await window.fdc3.joinChannel(channels[0].id);
-        return channels[0];
+  //get broadcast service and broadcast the given context type
+  async broadcastContextItem(contextType, channel, historyItems) {
+    let broadcastService = this.getBroadcastService(channel.type);
+    await broadcastService.broadcast(contextType, historyItems, channel);
+  }
+
+  //get app/system channel broadcast service
+  getBroadcastService(currentChannelType) {
+    if (currentChannelType === channelType.system) {
+      return this.systemChannelBroadcastService;
+    } else if (currentChannelType === channelType.app) {
+      return this.appChannelBroadcastService;
     }
+  }
+
+  //app channel broadcast service
+  appChannelBroadcastService = {
+    broadcast: async (contextType, historyItems, channel) => {
+      if (channel !== undefined) {
+        for (let i = 0; i < historyItems; i++) {
+          await channel.broadcast({
+            type: contextType,
+            name: `History-item-${i + 1}`,
+          });
+        }
+      }
+    },
+  };
 
-    async broadcast(contextType, historyItem) {
+  //system channel broadcast service
+  systemChannelBroadcastService = {
+    broadcast: async (contextType, historyItems) => {
+      for (let i = 0; i < historyItems; i++) {
         await window.fdc3.broadcast({
-            "type": contextType,
-            "name": `History-item-${historyItem}`
+          type: contextType,
+          name: `History-item-${i + 1}`,
         });
-    }
+      }
+    },
+  };
 
-    async addContextListener(contextType) {
-        await window.fdc3.addContextListener(
-            contextType,
-            () => closeFinsembleWindow());
+  //close ChannelsApp on completion and respond to app A
+  async closeWindowOnCompletion(channel, config) {
+    if (channel.type === channelType.system) {
+      await window.fdc3.addContextListener("closeWindow", async () => {
+        window.close();
+        await window.fdc3.broadcast({type: "windowClosed"});
```

Comment:
this may be unreachable... Some DAs may tweak window closing (as we do - its possible to remotely delay or cancel a close), others may not. Hence, I think you need to broadcast it before closing. 

If you have any trouble with closing in Finsemble let me know. Setting `manifest.foreign.components.Window Manager.titlebarType = "injected"` in the affected components config will resolve if you do.

Reply:
`window.close` doesn't stop the code that follows from running. I tested this theory out in an online javascript playground before implementing this. Also, the `windowClosed` listener that is in the `afterEach` in fdc3.broadcasts.ts is receiving the `windowClosed` context as it should and the Finsemble windows close as expected.